core/src/test/java/org/elasticsearch/transport/netty/NettyTransportPublishAddressIT.java

```diff
@@ -0,0 +1,72 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.netty;
+
+import org.elasticsearch.action.admin.cluster.node.info.NodeInfo;
+import org.elasticsearch.action.admin.cluster.node.info.NodesInfoResponse;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.transport.BoundTransportAddress;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.transport.TransportModule;
+
+import static org.hamcrest.Matchers.equalTo;
+
+/**
+ * Checks that Elasticsearch produces a sane publish_address when it binds to
+ * different ports on ipv4 and ipv6.
+ */
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class NettyTransportPublishAddressIT extends ESIntegTestCase {
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder()
+                .put(super.nodeSettings(nodeOrdinal))
+                .put(TransportModule.TRANSPORT_TYPE_KEY, "netty")
+                .put("node.mode", "network").build();
+    }
+
+    public void testDifferentPorts() throws Exception {
+        logger.info("--> starting a node on ipv4 only");
+        Settings ipv4Settings = Settings.builder().put("network.host", "127.0.0.1").build();
+        internalCluster().startNode(ipv4Settings); // should bind 127.0.0.1:9400
+
+        logger.info("--> starting a node on ipv4 and ipv6");
+        Settings bothSettings = Settings.builder().put("network.host", "_local_").build();
+        internalCluster().startNode(bothSettings); // should bind [::1]:9400 and 127.0.0.1:9401
+
+        logger.info("--> waiting for the cluster to declare itself stable");
+        ensureStableCluster(2); // fails if port of publish address does not match corresponding bound address
+
+        logger.info("--> checking if boundAddress matching publishAddress has same port");
+        NodesInfoResponse nodesInfoResponse = client().admin().cluster().prepareNodesInfo().get();
+        for (NodeInfo nodeInfo : nodesInfoResponse) {
+            BoundTransportAddress boundTransportAddress = nodeInfo.getTransport().getAddress();
+            boolean foundMatch = false;
+            for (TransportAddress boundAddress : boundTransportAddress.boundAddresses()) {
+                if (boundTransportAddress.publishAddress().sameHost(boundAddress)) {
+                    foundMatch = true;
+                    assertThat(boundAddress.getPort(), equalTo(boundTransportAddress.publishAddress().getPort()));
+                }
+            }
+            assertTrue("Could not find a boundAddress matching publishAddress", foundMatch);
+        }
```

Comment:
I changed the assertions in the test to make them closer to the original test you showed me. However, some assertions cannot be added without making the test flaky. For example, we cannot guarantee that port of ipv6 and ipv4 for `bothNode` is different (As there are other integration tests binding to ports as well).